### PR TITLE
rasdaemon: init at 0.6.4

### LINKED ISSUE #42592: Package request: Rasdaemon
## Issue description

`mcelog` is available in `nixpkgs`, but the utility has been deprecated in favor of `rasdaemon`. In particular, Rasdaemon has better support for AMD cards. No `rasdaemon` package currently exists in `nixpkgs`.

There is a Arch Linux PKGBUILD [here](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=rasdaemon).

Unfortunately, performing a manual install did not go well for me:
```
sudo make install
Making install in libtrace
make[1]: Entering directory '/home/matt/Downloads/rasdaemon-0.6.0/libtrace'
  AR       libtrace.a
make[1]: *** [Makefile:399: libtrace.a] Error 1
make[1]: Leaving directory '/home/matt/Downloads/rasdaemon-0.6.0/libtrace'
make: *** [Makefile:677: install-recursive] Error 1
```

### PR DESCRIPTION
**Pull request by leocp1**
Closes: #42592

###### Motivation for this change
new package.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/os-specific/linux/rasdaemon/default.nix [hunk 1]
```diff
@@ -0,0 +1,102 @@
+{ stdenv, lib, fetchurl, automake, autoreconfHook, coreutils, dmidecode
+, glibcLocales, hwdata, kmod, libtool, makeWrapper, sqlite, perl, doCheck ? true
+}:
+stdenv.mkDerivation rec {
+  name = "rasdaemon";
+  version = "0.6.4";
+  src = fetchurl {
+    url =
+      "http://www.infradead.org/~mchehab/rasdaemon/rasdaemon-${version}.tar.bz2";
```
Comment (JohnAZoidberg) [inline pkgs/os-specific/linux/rasdaemon/default.nix]:
```suggestion
      "https://www.infradead.org/~mchehab/rasdaemon/rasdaemon-${version}.tar.bz2";
```

Why not get it on pagure? There they already have version `0.6.4-2`

File: pkgs/os-specific/linux/rasdaemon/default.nix [hunk 2]
```diff
@@ -0,0 +1,102 @@
+{ stdenv, lib, fetchurl, automake, autoreconfHook, coreutils, dmidecode
+, glibcLocales, hwdata, kmod, libtool, makeWrapper, sqlite, perl, doCheck ? true
+}:
+stdenv.mkDerivation rec {
+  name = "rasdaemon";
```
Comment (JohnAZoidberg) [inline pkgs/os-specific/linux/rasdaemon/default.nix]:
```suggestion
  pname = "rasdaemon";
```

File: pkgs/os-specific/linux/rasdaemon/default.nix [hunk 3]
```diff
@@ -0,0 +1,102 @@
+{ stdenv, lib, fetchurl, automake, autoreconfHook, coreutils, dmidecode
+, glibcLocales, hwdata, kmod, libtool, makeWrapper, sqlite, perl, doCheck ? true
```
Comment (JohnAZoidberg) [inline pkgs/os-specific/linux/rasdaemon/default.nix]:
```suggestion
, glibcLocales, hwdata, kmod, libtool, makeWrapper, sqlite, perl
```

It's not necessary, can be overridden with `.overrideAttr`.

File: pkgs/os-specific/linux/rasdaemon/default.nix [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  # Since these programs need permissions to these directories
+  # to run properly, we wrap the executables to create
+  # these directories on startup. (see postFixup)
+  postConfigure = ''
+    sed -e '/$(install_sh) -d "$(DESTDIR)\/var\/lib\/rasdaemon"/d' \
+    -e '/$(install_sh) -d "$(DESTDIR)\/etc\/ras\/dimm_labels.d"/d' \
+    -i Makefile
+  '';
+
+  inherit doCheck;
```
Comment (JohnAZoidberg) [inline pkgs/os-specific/linux/rasdaemon/default.nix]:
```suggestion
  doCheck = true;
```

File: pkgs/os-specific/linux/rasdaemon/default.nix [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+  installFlags = [
+    "sysconfdir=${placeholder "out"}/etc"
+    "localstatedir=${placeholder "out"}/var"
+  ];
+
+  postInstall = ''
+    mkdir -p $out/lib/systemd/system
```
Comment (JohnAZoidberg) [inline pkgs/os-specific/linux/rasdaemon/default.nix]:
```suggestion
```

Shouldn't be necessary with `install`.

File: pkgs/os-specific/linux/rasdaemon/default.nix [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    install -D -p -m 0644 misc/rasdaemon.service \
+      $out/lib/systemd/system/rasdaemon.service
+    install -D -p -m 0644 misc/ras-mc-ctl.service \
+      $out/lib/systemd/system/ras-mc-ctl.service
+  '';
+
+  postFixup = ''
+    # fix dmidecode and modprobe paths
```
Comment (JohnAZoidberg) [inline pkgs/os-specific/linux/rasdaemon/default.nix]:
```suggestion
    # Fix dmidecode and modprobe paths
```

Nitpick: Please start all your comment with a capital letter, just looks nicer :)

File: pkgs/os-specific/linux/rasdaemon/default.nix [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    sed -e 's|find_prog ("dmidecode")|"${dmidecode}/bin/dmidecode"|g' \
```
Comment (JohnAZoidberg) [inline pkgs/os-specific/linux/rasdaemon/default.nix]:
```suggestion
    substituteInPlace $out/sbin/ras-mc-ctl \
      --replace 'find_prog ("dmidecode")' '${dmidecode}/bin/dmidecode"' \
      --replace 'find_prog ("modprobe")  or exit (1)' '${kmod}/bin/modprobe'
```

`substituteInPlace` prints a warning if the search string isn't found anymore.

File: nixos/modules/hardware/rasdaemon.nix [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+        '';
+      };
+    };
+  };
+
+  config = mkIf config.hardware.rasdaemon.enable {
+    systemd = {
+      packages = [ pkgs.rasdaemon ];
+
+      services.rasdaemon.wantedBy = [ "multi-user.target" ];
```
Comment (JohnAZoidberg) [inline nixos/modules/hardware/rasdaemon.nix]: Is that necessary? The upstream service files already contain this.

File: nixos/modules/hardware/rasdaemon.nix [hunk 9]
```diff
@@ -0,0 +1,26 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+{
+  options = {
+    hardware.rasdaemon = {
+      enable = mkOption {
```
Comment (JohnAZoidberg) [inline nixos/modules/hardware/rasdaemon.nix]:
```suggestion
      enable = mkEnableOption "RAS logging daemon"
```

Get's rid of all of the boilerplate below :)

File: pkgs/os-specific/linux/rasdaemon/default.nix [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+    coreutils
+    dmidecode
+    glibcLocales
+    hwdata
+    kmod
+    sqlite
+    (perl.withPackages (ps: with ps; [ DBI DBDSQLite ]))
+  ];
+
+  dontAddDisableDepTrack = true;
```
Comment (JohnAZoidberg) [inline pkgs/os-specific/linux/rasdaemon/default.nix]: Why?

File: pkgs/os-specific/linux/rasdaemon/default.nix [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+    wrapProgram $out/sbin/ras-mc-ctl \
+      --run "${coreutils}/bin/mkdir -p /etc/ras/dimm_labels.d"
+    # patch in LOCALE_ARCHIVE into systemd service
+    s='Environment="LOCALE_ARCHIVE=${glibcLocales}\/lib\/locale\/locale-archive"'
+    sed -e '/\[Service\]/a '"$s" \
+      -i $out/lib/systemd/system/ras-mc-ctl.service
+  '';
+
+  meta = with stdenv.lib; {
+    description = ''
```
Comment (JohnAZoidberg) [inline pkgs/os-specific/linux/rasdaemon/default.nix]:
```suggestion
    description = "A Reliability, Availability and Serviceability (RAS) logging tool using EDAC kernel tracing events"
```

I can't find the styleguide right now, but `description` should be short and more of this style.
You can put a longer description in `longDescription`. Maybe the one from [GitHub](https://github.com/mchehab/rasdaemon):
> Rasdaemon is a RAS (Reliability, Availability and Serviceability) logging tool. It records memory errors, using the EDAC tracing events. EDAC is a Linux kernel subsystem with handles detection of ECC errors from memory controllers for most chipsets on i386 and x86_64 architectures. EDAC drivers for other architectures like arm also exists.

File: pkgs/os-specific/linux/rasdaemon/default.nix [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+    sed -e 's|find_prog ("dmidecode")|"${dmidecode}/bin/dmidecode"|g' \
+    -e 's|find_prog ("modprobe")  or exit (1)|"${kmod}/bin/modprobe"|g' \
+    -i $out/sbin/ras-mc-ctl
+    # create database and label dirs on startup
+    wrapProgram $out/sbin/rasdaemon \
+      --run "${coreutils}/bin/mkdir -p /var/lib/rasdaemon"
+    wrapProgram $out/sbin/ras-mc-ctl \
+      --run "${coreutils}/bin/mkdir -p /etc/ras/dimm_labels.d"
+    # patch in LOCALE_ARCHIVE into systemd service
+    s='Environment="LOCALE_ARCHIVE=${glibcLocales}\/lib\/locale\/locale-archive"'
```
Comment (JohnAZoidberg) [inline pkgs/os-specific/linux/rasdaemon/default.nix]: Use `|` instead of `/` for the sed delimiter. Then you don't need so much escaping :)

**Review by JohnAZoidberg — CHANGES_REQUESTED** [review]
Thanks a lot for your contribution! Welcome to nixpkgs.

How have you tested this program?

**Comment by leocp1** [comment]
Thank you for your suggestions! I have pushed most of them.

> How have you tested this program?

I manually ran the `rasdaemon` and `ras-mc-ctl` programs with every command line option, and checked that they seemed to run successfully.

There are tests under TESTING in rasdaemon's README (https://github.com/mchehab/rasdaemon/blob/v0.6.4/README) that should probably be in `nixos/tests`, but I have not added them.

Are these tests required? (I believe the similar `mcelog` package is not tested either).
If these tests are required, should the `contrib` scripts install to a split `dev` output? Should the `mce-inject`/`mce-test`/`aer-inject` programs be separated into their own packages?

> Why not get it on pagure? There they already have version 0.6.4-2

* Pagure does not have any release tarballs: https://releases.pagure.org/rasdaemon
* The latest commit on the main repository
  (http://git.infradead.org/users/mchehab/rasdaemon.git) is 0.6.4.
  The tag debian/0.6.4-2 on pagure just changes debian packaging files.

Another alternative to consider would be to use the github release, since the official tarball from infradead does not contain the `contrib` folder mentioned in the TESTING section.

> Use | instead of / for the sed delimiter. Then you don't need so much escaping :)

Sed delimiter replacement only applies to the 's' command. :(
(http://www.gnu.org/software/sed/manual/sed.html#The-_0022s_0022-Command)

**Comment by Evils-Devils** [comment]
i've had a go at cleaning this up a bit [here](https://github.com/Evils-Devils/nixpkgs/tree/rasdaemon)
(i'm not sure it's a good practice to have `maintainer-list.nix` entry authored by someone that is not the maintainer being added but this should be a separate commit)
handling `/etc/ras/dimm_labels.d` with the module is a bit tricky without expected content so i've left that as a TODO
maybe we can find someone with a supported Dell machine for that?
testing the module would probably require compiling a custom kernel and packaging of the error injection tools, that's probably worth it, but a bit more work/research than i'm currently willing to put in

**Comment by evils** [comment]
i'm closing this as my #85039 is based on this and if not improved, at least more up to date

@leocp1 feel free to reopen this, help out with my PR, or let me know if you'd like to be added as a maintainer on that

and thanks for providing a solid start

**Pull request closed without merging**